contract/src/contract.rs

```diff
@@ -95,24 +177,130 @@ pub fn update_config(
+            }
+            messages.push(CosmosMsg::from(BankMsg::Burn {
+                amount: info.funds.clone(),
+            }));
+        }
+    }
+    // Big deviation here lets break it down, end result we want the amount burned and all the messages ready
+    // If using cw20 (first case) we will need to use the CW20 spec and rely on the token address being stored in Config
+    // Otherwise we are proceeding as normal using Native tokens and minting
+    let (amount, ash_amount) = match config.use_cw20 {
+        true => {
+            let provided_amount = cw20_msg.unwrap().amount;
+
+            // But dont return, after this we are going to mint the amount to the sender
+            if config.fee_collector_addr != Addr::unchecked("") {
+                let fee = provided_amount * config.burn_fee;
+
+                messages.push(build_transfer_cw20_token_msg(
+                    config.fee_collector_addr.clone(),
+                    config.mint_denom.clone(),
+                    fee,
+                )?);
+            }
+
+            let ash_amount = coins(provided_amount.u128(), config.mint_denom.as_str());
+
+            // Now mint the ash to the sender, first do a CW20 Mint, this contract already has minter rights
+            messages.push(build_mint_cw20_token_msg(
+                env.contract.address,
+                config.mint_denom.clone(),
+                provided_amount,
+            )?);
+            // Now send ash to the sender
+            messages.push(build_transfer_cw20_token_msg(
+                info.sender.clone(),
+                config.mint_denom.clone(),
+                provided_amount,
+            )?);
+            (provided_amount, ash_amount)
+        }
+        false => {
+            if info.funds.len() != 1 {
+                return Err(ContractError::IncorrectTokenQuantity {});
+            }
+            //TODO: Remove or force user to set the native token denom on instantiate
+            if info.funds[0].denom != "uwhale" {
+                return Err(ContractError::IncorrectToken {});
+            }
+
+            let amount = info.funds[0].amount;
+
+            // If the fee_collect_addr is set, inflate the amount to mint, send the difference (fee) to the fee_collector_addr
+            // But dont return, after this we are going to mint the amount to the sender
+            if config.fee_collector_addr != Addr::unchecked("") {
+                let fee = amount * config.burn_fee;
+                let fee_amount = coins(fee.u128(), config.mint_denom.as_str());
+                messages.push(CosmosMsg::Bank(BankMsg::Send {
+                    to_address: config.fee_collector_addr.to_string(),
+                    amount: fee_amount.clone(),
+                }));
+            }
+
+            //mint ASH and transfer to sender
+            messages.push(<MsgMint as Into<CosmosMsg>>::into(MsgMint {
+                sender: env.contract.address.to_string(),
```

Comment:
Here it's correct, although i think in a newer version of the token factory we can mint directly to the recepient address, but let's go with this implementation for now.